Show 'Editable Children' and 'Load as Placeholder' context menu options for multi-node selections

tom-jk opened the pull request:
Partially addresses godotengine/godot-proposals#581.

Makes 'Editable Children' and ‘Load as Placeholder’ options available in the scene tree dock context menu for multi-node selections.

If multiple nodes are selected, some with the setting enabled and some without, the user is asked to make them all editable/placeholders, or all not.

![Screenshot_2023-10-18_17-31-37](https://github.com/godotengine/godot/assets/64640811/2425e38e-718a-472f-b6be-86ada2daf3f7)

If the selection contains any node for which an option would not be applicable, the option will not appear in the menu.

**Notes/Questions**

Context menus don't support indeterminate checkboxes, so only shows a check if all selected nodes are enabled.

User can be shown enable/disable placeholders prompt, then disable editable children prompt back-to-back. Is this ok?

tom-jk committed:
Show Editable, Placeholder context menu options for multi-node selection

Makes 'Editable Children' and ‘Load as Placeholder’ options available in
the scene tree dock context menu for multi-node selections.

If multiple nodes are selected, some with the setting enabled and some
without, the user is asked to make them all editable/placeholders, or
all not.

If the selection contains any node for which an option would not be
applicable, the option will not appear in the menu.

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -3159,6 +3248,26 @@ void SceneTreeDock::_tree_rmb(const Vector2 &p_menu_pos) {
+				has_internal |= node->get_scene_file_path().is_empty();
+				has_top_level |= node->get_owner() == nullptr;
+				has_non_placeholder |= !node->get_scene_instance_load_placeholder();
+				if (has_internal or has_top_level) {
```
```suggestion
				if (has_internal || has_top_level) {
```
No alternative syntax

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -3159,6 +3248,26 @@ void SceneTreeDock::_tree_rmb(const Vector2 &p_menu_pos) {
+				}
+			}
+		}
+		if (!has_internal and !has_top_level and profile_allow_editing) {
```
```suggestion
		if (!has_internal && !has_top_level && profile_allow_editing) {
```

tom-jk commented:
- Don't show options if child and ancestor selected (selection size != full_selection size), to avoid potentially confusing behaviour.
- move tree update out of loops. With a couple hundred nodes, operation time goes from >1 sec to immediate.

tom-jk commented:
- rebased.
 - change "Load as Placeholder" and "Editable Children" in dialogs to "use_placeholder" and "editable_instance" to match existing dialog.

tom-jk commented:
I rebased and fixed merge conflict locally, but I'm not sure now if I should go back to "Load as Placeholder" and "Editable Children" in dialogs? In master there's a dialog that says 

`Disabling "editable_instance" will cause all properties of the node to be reverted to their default.` 

and another that says 

`Enabling "Load as Placeholder" will disable "Editable Children" and cause all properties of the node to be reverted to their default.`

so it's not clear which example I should follow. (fwiw I think the presentation in the latter is better and presumably what's intended.)

tom-jk commented:
Rebased, and decided on `"Load as Placeholder"` and `"Editable Children"`.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -4093,10 +4271,24 @@ SceneTreeDock::SceneTreeDock(Node *p_scene_root, EditorSelection *p_editor_selec
 	editable_instance_remove_dialog->connect("confirmed", callable_mp(this, &SceneTreeDock::_toggle_editable_children_from_selection));
 
+	placeholder_enable_or_disable_dialog = memnew(ConfirmationDialog);
+	placeholder_enable_or_disable_dialog->set_ok_button_text("Enable for all");
```
```suggestion
	placeholder_enable_or_disable_dialog->set_ok_button_text(TTR("Enable for all"));
```
Same for all buttons.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -1023,51 +1023,119 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
+					if (node) {
+						bool editable = EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(node);
+						has_editables |= editable;
+						has_non_editables |= !editable;
```
You could break if both are true.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -1023,51 +1023,119 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
+
+				// Fire dialog if selection does contain both, to ask user what they want to do (enable or disable).
+				if (has_non_editables && has_editables) {
+					editable_instance_enable_or_disable_dialog->set_text(TTR("Selection contains some nodes that have \"Editable Children\" enabled and some that do not."));
```
This is only done here, so should be moved to constructor.

tom-jk replied:
also done for `placeholder_enable_or_disable_dialog`

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -1023,51 +1023,119 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
+				// check if the selection contains both editable and non-editable nodes.
+				bool has_editables = false;
+				bool has_non_editables = false;
+				for (int i = 0; i < selection.size(); i++) {
```
```suggestion
				for (const Node *node : selection) {
```
Same below (but without `const` probably).

KoBeWi commented on editor/scene_tree_dock.h:
```diff
@@ -152,9 +152,16 @@ class SceneTreeDock : public VBoxContainer {
+	Button *placeholder_disable_all_button = nullptr;
 	ConfirmationDialog *placeholder_editable_instance_remove_dialog = nullptr;
 
+	bool editable_instance_change_operation_is_enable = false;
+	bool placeholder_change_operation_is_enable = false;
```
```suggestion
	bool editable_instance_change_operation_is_enabled = false;
	bool placeholder_change_operation_is_enabled = false;
```

tom-jk replied:
These I use to temporarily store the result of one dialog if there might be another dialog before doing the operation, so the meaning is less "a something is enabled" and more "the thing we are about to do is enable". 

The suggested change to "enabled" might imply that nothing will happen if it is false (e.g. "the operation is not enabled"), when it instead means that the user chose to disable all rather than enable all.

How about `editable_instance_enable_or_disable_dialog_result_was_enable` and `placeholder_enable_or_disable_dialog_result_was_enable` ? Bit of a mouthful, but purpose more explicit.

Or, remove the `_was_enable` and use an enum `{DISABLE_ALL, ENABLE_ALL}` instead?

Actually, since only one of these will be in play at any given time, could just roll into a single `dialog_result_was_enable` bool?

KoBeWi replied:
Yep, I think they should be unified. Maybe something like `dialog_option_enable_all`? Though "dialog" is a bit ambiguous, but I don't have better idea (that isn't super long).

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -1023,51 +1023,119 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
-						editable_instance_remove_dialog->popup_centered();
-						break;
+			if (selection.size() > 1) {
+				// check if the selection contains both editable and non-editable nodes.
```
```suggestion
				// Check if the selection contains both editable and non-editable nodes.
```

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+		}
+	} else {
+		for (Node *node : selection) {
+			if (node) {
```
```suggestion
			if (!node) {
				continue;
			}
```
Less indentation.

I think you could also remove the upper condition; range loop for a single element does not really make difference.

tom-jk replied:
I remember I added the `if (node)` checks when I was just getting started and saw it being used elsewhere, but I have since wondered if they're necessary at all?[^1] I grep'd through the codebase and only really saw such checks being used in situations where it was clearly sensible to do so, while my use of it seems overly paranoid.

[^1]: I figured maybe it guards against a node being freed in the meantime. I removed the checks and tested with a tool node that can free itself while the rmb menu is open; because doing so also removes it from the scene tree dock and therefore the selection, it causes no issues. Seems the selection and the scene would have to diverge in the middle of the operation, or the selection array be fouled up some other way, both of which seem improbable. Though I think I now speculate on things outside my understanding...

tom-jk replied:
Re: removing the condition for single-node selection: it probably wouldn't save much space vertically as is (though would further reduce indentation). We would still need to catch the single-node case in order to set `dialog_option_enable_all = !editable`, before falling through to the for-loop. Like this:

```
	if (selection.size() == 1) {
		List<Node *>::Element *e = selection.front();
		if (e) {
			dialog_option_enable_all = !EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(e->get());
		}
	}
	for (Node *node : selection) {
		if (!node) {
			continue;
		}
		bool editable = !EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(node);

		if (editable == dialog_option_enable_all) {
			_toggle_editable_children(node);
		}
	}
```

Maybe I should just add a `_set_editable_children(node, editable)` method? Something roughly like this:

```
	for (Node *node : selection) {
		if (!node) {
			continue;
		}
		_set_editable_children(node, dialog_option_enable_all);
	}

...

void SceneTreeDock::_set_editable_children(Node *p_node, bool p_editable) {
	if (p_node) {
		bool editable = EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(p_node);
		if (!editable == p_editable) {
			// set here or call _toggle_editable_children
			// (if set here, maybe make _toggle_editable_children call here instead to avoid duplication)
		}
	}
}
```

KoBeWi replied:
So `dialog_option_enable_all` is not initialized for single node? Then I guess it's not worth to change it.

KoBeWi reviewed (dismissed):
Left some minor comments, but otherwise looks fine (though not a fan of the added code complexity .-.).

Note that this PR conflicts with #60974, and it will need to be updated if that PR is merged first.

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+	} else {
+		for (Node *node : selection) {
+			if (node) {
+				if (dialog_option_enable_all == true) {
```
```suggestion
				if (dialog_option_enable_all) {
```

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+				if (editable) {
+					_toggle_editable_children(node);
+
+					bool placeholder = node->get_scene_instance_load_placeholder();
```
```suggestion
					bool placeholder = !node->get_scene_instance_load_placeholder();
```

KoBeWi replied:
This suggestion is confusing alone. You can make comments for multiple lines.

AThousandShips replied:
I know I was on phone, I will update it

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+					_toggle_editable_children(node);
+
+					bool placeholder = node->get_scene_instance_load_placeholder();
+					placeholder = !placeholder;
```
```suggestion
```

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+					}
+				}
+
+				bool placeholder = node->get_scene_instance_load_placeholder();
```
```suggestion
				bool placeholder = !node->get_scene_instance_load_placeholder();
```

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+				}
+
+				bool placeholder = node->get_scene_instance_load_placeholder();
+				placeholder = !placeholder;
```
```suggestion
```

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+					}
+				}
+
+				bool placeholder = node->get_scene_instance_load_placeholder();
+				placeholder = !placeholder;
```
```suggestion
				bool placeholder = !node->get_scene_instance_load_placeholder();
```

KoBeWi replied:
Actually this can be simplified to
```C++
bool placeholder = node->get_scene_instance_load_placeholder();
if (placeholder != dialog_option_enable_all) {
```
I don't get why you are inverting bool and checking for equality.

AThousandShips replied:
Because this line:
```
node->set_scene_instance_load_placeholder(placeholder);
```

That one would have to be inverted instead

KoBeWi replied:
Then I think it should be `node->set_scene_instance_load_placeholder(dialog_option_enable_all)`?

AThousandShips replied:
I think it might be more confusing than the current code, but that would work

tom-jk replied:
ok, I now have

```
bool placeholder = node->get_scene_instance_load_placeholder();

if (placeholder != dialog_option_enable_all) {
	node->set_scene_instance_load_placeholder(!placeholder);
}
```
the `if` is so as to not set placeholder to the same value it already is - I guess I thought it might avoid godot doing unnecessary work when the value didn't change. I think this reasoning was flawed, so probably good to just remove the `if`?

```
bool placeholder = node->get_scene_instance_load_placeholder();
node->set_scene_instance_load_placeholder(!placeholder);
```

KoBeWi replied:
Well the `set_scene_instance_load_placeholder()` code is so minimal that the condition does matter indeed.

tom-jk replied:
realised that `if` was also making it work like a toggle. correcting my last comment, it can now just be one line:

```
node->set_scene_instance_load_placeholder(dialog_option_enable_all);
```

AThousandShips commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+				if (editable) {
+					_toggle_editable_children(node);
+
+					bool placeholder = node->get_scene_instance_load_placeholder();
+					placeholder = !placeholder;
```
```suggestion
					bool placeholder = !node->get_scene_instance_load_placeholder();
```

tom-jk replied:
I'll do likewise for lines 2359-2360: `bool editable = ( ... ); editable = !editable;`

AThousandShips replied:
Good knew I had seen one more that I had missed!

tom-jk replied:
pushed the following:

```
bool placeholder = node->get_scene_instance_load_placeholder();
node->set_scene_instance_load_placeholder(!placeholder);
```

My thinking was that `placeholder` represent the placeholder state as it currently is, which we will then negate to get the new one. In your suggested change, `placeholder` represents the new value, the future/intended state. I suppose one could argue for either, and so I probably should have asked before pushing (sorry): which form you think would be the more proper?

tom-jk commented on editor/scene_tree_dock.cpp:
```diff
@@ -2263,31 +2331,114 @@ void SceneTreeDock::_shader_creation_closed() {
+			Node *node = e->get();
+			if (node) {
+				bool editable = EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(node);
+				if (editable) {
+					_toggle_editable_children(node);
+
+					bool placeholder = node->get_scene_instance_load_placeholder();
+					placeholder = !placeholder;
+
+					node->set_scene_instance_load_placeholder(placeholder);
+				}
```
setting placeholder shouldn't be inside the `if (editable)` block.
this code is never reached in the case of a single-node selection that is not editable (we only come here if we needed to ask user a question first) so correcting it won't make any functional difference, but will do so anyway.

tom-jk commented:
Rebased.

Calinou commented on editor/scene_tree_dock.cpp:
```diff
@@ -4132,10 +4301,26 @@ SceneTreeDock::SceneTreeDock(Node *p_scene_root, EditorSelection *p_editor_selec
 
+	editable_instance_enable_or_disable_dialog = memnew(ConfirmationDialog);
+	editable_instance_enable_or_disable_dialog->set_text(TTR("Selection contains some nodes that have \"Editable Children\" enabled and some that do not."));
+	editable_instance_enable_or_disable_dialog->set_ok_button_text(TTR("Enable for all"));
```
Use Title Case for buttons:

```suggestion
	editable_instance_enable_or_disable_dialog->set_ok_button_text(TTR("Enable for All"));
```

Calinou commented on editor/scene_tree_dock.cpp:
```diff
@@ -4132,10 +4301,26 @@ SceneTreeDock::SceneTreeDock(Node *p_scene_root, EditorSelection *p_editor_selec
 
+	placeholder_enable_or_disable_dialog = memnew(ConfirmationDialog);
+	placeholder_enable_or_disable_dialog->set_text(TTR("Selection contains some nodes that have \"Load as Placeholder\" enabled and some that do not."));
+	placeholder_enable_or_disable_dialog->set_ok_button_text(TTR("Enable for all"));
```
```suggestion
	placeholder_enable_or_disable_dialog->set_ok_button_text(TTR("Enable for All"));
```

Calinou commented on editor/scene_tree_dock.cpp:
```diff
@@ -4132,10 +4301,26 @@ SceneTreeDock::SceneTreeDock(Node *p_scene_root, EditorSelection *p_editor_selec
+	placeholder_enable_or_disable_dialog = memnew(ConfirmationDialog);
+	placeholder_enable_or_disable_dialog->set_text(TTR("Selection contains some nodes that have \"Load as Placeholder\" enabled and some that do not."));
+	placeholder_enable_or_disable_dialog->set_ok_button_text(TTR("Enable for all"));
+	placeholder_disable_all_button = placeholder_enable_or_disable_dialog->add_button(TTR("Disable for all"), true);
```
```suggestion
	placeholder_disable_all_button = placeholder_enable_or_disable_dialog->add_button(TTR("Disable for All"), true);
```

Calinou commented on editor/scene_tree_dock.cpp:
```diff
@@ -4132,10 +4301,26 @@ SceneTreeDock::SceneTreeDock(Node *p_scene_root, EditorSelection *p_editor_selec
+	editable_instance_enable_or_disable_dialog = memnew(ConfirmationDialog);
+	editable_instance_enable_or_disable_dialog->set_text(TTR("Selection contains some nodes that have \"Editable Children\" enabled and some that do not."));
+	editable_instance_enable_or_disable_dialog->set_ok_button_text(TTR("Enable for all"));
+	editable_instance_disable_all_button = editable_instance_enable_or_disable_dialog->add_button(TTR("Disable for all"), true);
```
```suggestion
	editable_instance_disable_all_button = editable_instance_enable_or_disable_dialog->add_button(TTR("Disable for All"), true);
```

tom-jk commented:
Title case for buttons.

YuriSizov commented on editor/scene_tree_dock.cpp:
```diff
@@ -3264,6 +3408,31 @@ void SceneTreeDock::_tree_rmb(const Vector2 &p_menu_pos) {
+					}
+				}
+			}
+			if (!has_internal && !has_top_level && profile_allow_editing) {
```
`profile_allow_editing` can be checked before going through all the trouble with checking the selection.

YuriSizov commented on editor/scene_tree_dock.cpp:
```diff
@@ -1032,51 +1032,119 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
+
+				if (has_editables) {
+					dialog_option_enable_all = false;
+					editable_instance_remove_dialog->set_text(TTR("Disabling \"Editable Children\" will cause all properties of those nodes to be reverted to their default."));
```
```suggestion
					editable_instance_remove_dialog->set_text(TTR("Disabling \"Editable Children\" will cause all properties of selected nodes to be reverted to their default."));
```

Seems like you've copied this text from another context where it was clear which are "those" nodes, but here it isn't.

YuriSizov commented on editor/scene_tree_dock.cpp:
```diff
@@ -2271,31 +2339,109 @@ void SceneTreeDock::_shader_creation_closed() {
+	dialog_option_enable_all = false;
+
+	// Fire confirmation dialog.
+	editable_instance_remove_dialog->set_text(TTR("Disabling \"Editable Children\" will cause all properties of those nodes to be reverted to their default."));
```
```suggestion
	editable_instance_remove_dialog->set_text(TTR("Disabling \"Editable Children\" will cause all properties of selected nodes to be reverted to their default."));
```

Same here.

YuriSizov commented on editor/scene_tree_dock.cpp:
```diff
@@ -2271,31 +2339,109 @@ void SceneTreeDock::_shader_creation_closed() {
+			if (!node) {
+				return;
+			}
+			bool editable = EditorNode::get_singleton()->get_edited_scene()->is_editable_instance(node);
+			if (editable) {
+				_toggle_editable_children(node);
+			}
```
Shouldn't this be called only when we enable load as placeholder?

YuriSizov replied:
That's definitely ripe for refactoring already. But I guess we can live with it for the moment.

YuriSizov commented:
> though not a fan of the added code complexity .-.

I agree, some of these methods can be simplified and paths for single and multi selection can be mostly unified. It should also help to reduce the number of errors. That said, I think we can still merge it as is (once fixed up), and do a clean up later. Stuff like `_tool_selected` should probably be split into more manageable blocks anyway (and this is true for a lot of our codebase).

tom-jk commented:
'those' -> 'selected', earlier `profile_allow_editing` check.
